File: python/bsti2060/telemetry.py
from __future__ import annotations

import base64
import hashlib
import hmac
import json
from dataclasses import asdict, dataclass
from typing import Any, Literal

SourceNode = Literal["HIN", "AIMN"]
Stream = Literal["thermal_hydration", "ecg", "status", "isolation"]
Quality = Literal["valid", "degraded", "unavailable", "stale", "isolated"]

@dataclass(frozen=True)
class Integrity:
    algorithm: Literal["HMAC-SHA-256"]
    tag: str
    keyId: str

@dataclass(frozen=True)
class TelemetryEnvelope:
    version: Literal["1.0"]
    sourceNode: SourceNode
    sequence: int
    timestampMs: int
    stream: Stream
    quality: Quality
    payload: dict[str, Any]
    integrity: Integrity

def _body(envelope: TelemetryEnvelope | dict[str, Any]) -> bytes:
    data = asdict(envelope) if isinstance(envelope, TelemetryEnvelope) else envelope
    data = {k: v for k, v in data.items() if k != "integrity"}
    return json.dumps(data, separators=(",", ":"), ensure_ascii=False).encode()

def sign_envelope(body: dict[str, Any], key: bytes, key_id: str = "development-key") -> TelemetryEnvelope:
    tag = base64.urlsafe_b64encode(hmac.new(key, _body(body), hashlib.sha256).digest()).rstrip(b"=").decode()
    return TelemetryEnvelope(**body, integrity=Integrity("HMAC-SHA-256", tag, key_id))

def verify_envelope(envelope: TelemetryEnvelope, key: bytes) -> bool:
    expected = hmac.new(key, _body(envelope), hashlib.sha256).digest()
    padded = envelope.integrity.tag + "=" * (-len(envelope.integrity.tag) % 4)
    try:
        received = base64.urlsafe_b64decode(padded)
    except ValueError:
        return False
    return envelope.integrity.algorithm == "HMAC-SHA-256" and hmac.compare_digest(received, expected)

def assert_supported_rate(stream: Stream, rate_hz: int) -> None:
    expected = {"ecg": 250, "thermal_hydration": 10}.get(stream)
    if expected is not None and rate_hz != expected:
        raise ValueError(f"Unsupported {stream} polling rate: expected {expected} Hz, received {rate_hz} Hz")

File: typescript/src/telemetry.ts
import { createHmac, timingSafeEqual } from "node:crypto";

export type SourceNode = "HIN" | "AIMN";
export type Stream = "thermal_hydration" | "ecg" | "status" | "isolation";
export type Quality = "valid" | "degraded" | "unavailable" | "stale" | "isolated";

export interface TelemetryEnvelope {
  version: "1.0";
  sourceNode: SourceNode;
  sequence: number;
  timestampMs: number;
  stream: Stream;
  quality: Quality;
  payload: Record<string, unknown>;
  integrity: { algorithm: "HMAC-SHA-256"; tag: string; keyId: string };
}

export function canonicalBody(envelope: Omit<TelemetryEnvelope, "integrity">): string {
  return JSON.stringify(envelope);
}

export function signEnvelope(
  envelope: Omit<TelemetryEnvelope, "integrity">,
  key: Buffer,
  keyId = "development-key"
): TelemetryEnvelope {
  const tag = createHmac("sha256", key).update(canonicalBody(envelope)).digest("base64url");
  return { ...envelope, integrity: { algorithm: "HMAC-SHA-256", tag, keyId } };
}

export function verifyEnvelope(envelope: TelemetryEnvelope, key: Buffer): boolean {
  const { integrity, ...body } = envelope;
  if (integrity.algorithm !== "HMAC-SHA-256") return false;
  const expected = createHmac("sha256", key).update(canonicalBody(body)).digest();
  const received = Buffer.from(integrity.tag, "base64url");
  return received.length === expected.length && timingSafeEqual(received, expected);
}

export function assertSupportedRate(stream: Stream, rateHz: number): void {
  const expected = stream === "ecg" ? 250 : stream === "thermal_hydration" ? 10 : undefined;
  if (expected !== undefined && rateHz !== expected) {
    throw new Error(`Unsupported ${stream} polling rate: expected ${expected} Hz, received ${rateHz} Hz`);
  }
}
